circle/src/twiddles.rs

```diff
@@ -0,0 +1,77 @@
+use alloc::vec::Vec;
+use core::mem;
+
+use itertools::Itertools;
+use p3_field::batch_multiplicative_inverse;
+use p3_field::extension::{Complex, ComplexExtendable};
+use p3_util::linear_map::LinearMap;
+use tracing::instrument;
+
+use crate::domain::CircleDomain;
+
+#[derive(Default)]
+pub(crate) struct TwiddleCache<F>(
+    // (log_n, shift) -> (twiddles, inverse_twiddles)
+    #[allow(clippy::type_complexity)]
+    LinearMap<(usize, Complex<F>), (Vec<Vec<F>>, Option<Vec<Vec<F>>>)>,
+);
+
+impl<F: ComplexExtendable> TwiddleCache<F> {
+    pub(crate) fn get_twiddles(
+        &mut self,
+        log_n: usize,
+        shift: Complex<F>,
+        inv: bool,
+    ) -> &Vec<Vec<F>> {
+        let cache = self
+            .0
+            .get_or_insert_with((log_n, shift), || (compute_twiddles(log_n, shift), None));
+        if !inv {
+            &cache.0
+        } else {
+            cache.1.get_or_insert_with(|| {
+                cache
+                    .0
+                    .iter()
+                    .map(|xs| batch_multiplicative_inverse(xs))
+                    .collect()
+            })
+        }
+    }
+}
+
+#[instrument(skip(shift))]
+fn compute_twiddles<F: ComplexExtendable>(log_n: usize, shift: Complex<F>) -> Vec<Vec<F>> {
```

Comment:
Computes all (non-inverted) twiddles for the FFT over a circle domain of size `2^log_n`, for all layers of the FFT.